File: AnkiDroid/src/test/java/com/ichi2/anki/AddonModelTest.java
package com.ichi2.anki;

import com.ichi2.utils.JSONObject;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import androidx.test.ext.junit.runners.AndroidJUnit4;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;
import static org.hamcrest.Matchers.nullValue;

@RunWith(AndroidJUnit4.class)
public class AddonModelTest extends RobolectricTest {
    private static final String progrssBarAddonPackageJson = "{\n" +
            "  \"name\": \"ankidroid-js-addon-progress-bar\",\n" +
            "  \"version\": \"1.0.5\",\n" +
            "  \"description\": \"Show progress bar in AnkiDroid, this package may not be used in node_modules. For using this addon view. https://github.com/ankidroid/Anki-Android/pull/7958\",\n" +
            "  \"main\": \"index.js\",\n" +
            "  \"ankidroid_js_api\": \"0.0.1\",\n" +
            "  \"addon_type\": \"reviewer\",\n" +
            "  \"scripts\": {\n" +
            "    \"test\": \"echo \\\"Error: no test specified\\\" && exit 1\"\n" +
            "  },\n" +
            "  \"repository\": {\n" +
            "    \"type\": \"git\",\n" +
            "    \"url\": \"git+https://github.com/infinyte7/Anki-Custom-Card-Layout.git\"\n" +
            "  },\n" +
            "  \"keywords\": [\n" +
            "    \"ankidroid-js-addon\"\n" +
            "  ],\n" +
            "  \"author\": \"infinyte7\",\n" +
            "  \"license\": \"MIT\",\n" +
            "  \"bugs\": {\n" +
            "    \"url\": \"https://github.com/infinyte7/Anki-Custom-Card-Layout/issues\"\n" +
            "  },\n" +
            "  \"homepage\": \"https://github.com/infinyte7/Anki-Custom-Card-Layout#readme\"\n" +
            "}\n";

    /*
      keywords not present
      "keywords": [
        "ankidroid-js-addon"
      ]
     */
    private static final String keywordsEmptyAddonPackageJson =  "{\n" +
            "  \"name\": \"ankidroid-js-addon-test-addon-1\",\n" +
            "  \"version\": \"1.0.1\",\n" +
            "  \"description\": \"Keywords not present\",\n" +
            "  \"main\": \"index.js\",\n" +
            "  \"ankidroid_js_api\": \"0.0.1\",\n" +
            "  \"addon_type\": \"reviewer\",\n" +
            "  \"scripts\": {\n" +
            "    \"test\": \"echo \\\"Error: no test specified\\\" && exit 1\"\n" +
            "  },\n" +
            "  \"repository\": {\n" +
            "    \"type\": \"git\",\n" +
            "    \"url\": \"git+https://github.com/infinyte7/ankidroid-js-addon.git\"\n" +
            "  },\n" +
            "  \"keywords\": [\n" +
            "  ],\n" +
            "  \"author\": \"Mani (infinyte7)\",\n" +
            "  \"license\": \"MIT\",\n" +
            "  \"bugs\": {\n" +
            "    \"url\": \"https://github.com/infinyte7/ankidroid-js-addon/issues\"\n" +
            "  },\n" +
            "  \"homepage\": \"https://github.com/infinyte7/ankidroid-js-addon#readme\"\n" +
            "}\n";

    /*
      all fields empty, only these fields checked for validity of addons
      name, version, ankidroid_js_api, addon_type, author, homepage
     */

    private static final String emptyAddonPackageJson = "{\n" +
            "  \"name\": \"\",\n" +
            "  \"version\": \"\",\n" +
            "  \"description\": \"\",\n" +
            "  \"main\": \"index.js\",\n" +
            "  \"ankidroid_js_api\": \"\",\n" +
            "  \"addon_type\": \"\",\n" +
            "  \"scripts\": {\n" +
            "    \"test\": \"echo \\\"Error: no test specified\\\" && exit 1\"\n" +
            "  },\n" +
            "  \"repository\": {\n" +
            "    \"type\": \"git\",\n" +
            "    \"url\": \"git+https://github.com/infinyte7/ankidroid-js-addon.git\"\n" +
            "  },\n" +
            "  \"keywords\": [\n" +
            "    \"ankidroid-js-addon\"\n" +
            "  ],\n" +
            "  \"author\": \"\",\n" +
            "  \"license\": \"MIT\",\n" +
            "  \"bugs\": {\n" +
            "    \"url\": \"https://github.com/infinyte7/ankidroid-js-addon/issues\"\n" +
            "  },\n" +
            "  \"homepage\": \"\"\n" +
            "}\n";

    private JSONObject progressBarJsonObject;
    private JSONObject emptyPackageJsonObject;
    private JSONObject keywordsEmptyJsonObject;

    @Before
    public void setUp() {
        progressBarJsonObject = new JSONObject(progrssBarAddonPackageJson);
        emptyPackageJsonObject = new JSONObject(emptyAddonPackageJson);
        keywordsEmptyJsonObject = new JSONObject(keywordsEmptyAddonPackageJson);
    }

    @Test
    public void isValidAddons() {
        AddonModel addonModel = AddonModel.tryParse(progressBarJsonObject, "reviewer");
        assertThat(addonModel.getName(), is("ankidroid-js-addon-progress-bar"));
        assertThat(addonModel.getVersion(), is("1.0.5"));
        assertThat(addonModel.getJsApiVersion(), is("0.0.1"));
        assertThat(addonModel.getType(), is("reviewer"));
        assertThat(addonModel.getDeveloper(), is("infinyte7"));
    }

    @Test
    public void testNullAddons() {
        AddonModel addonModel1 = AddonModel.tryParse(emptyPackageJsonObject, "reviewer");
        AddonModel addonModel2 = AddonModel.tryParse(keywordsEmptyJsonObject, "reviewer");

        assertThat(addonModel1, is(nullValue()));
        assertThat(addonModel2, is(nullValue()));
    }
}

File: AnkiDroid/src/test/java/com/ichi2/anki/jsaddons/AddonModelTest.kt
/*
 * Copyright (c) 2021 Mani infinyte01@gmail.com
 *
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 3 of the License, or (at your option) any later
 * version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 * PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.ichi2.anki.jsaddons

import androidx.test.ext.junit.runners.AndroidJUnit4
import com.ichi2.anki.AnkiDroidJsAPIConstants.CURRENT_JS_API_VERSION
import com.ichi2.anki.RobolectricTest
import com.ichi2.anki.jsaddons.AddonsConst.ANKIDROID_JS_ADDON_KEYWORDS
import com.ichi2.anki.jsaddons.AddonsConst.REVIEWER_ADDON
import com.ichi2.utils.FileOperation
import junit.framework.TestCase.assertEquals
import junit.framework.TestCase.assertFalse
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.core.StringEndsWith.endsWith
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import java.io.File
import java.io.IOException
import kotlin.test.assertIs

@RunWith(AndroidJUnit4::class)
class AddonModelTest : RobolectricTest() {
    private lateinit var validNpmPackageJson: File
    private lateinit var notValidNpmPackageJson: File
    private lateinit var addonsPackageListTestJson: String

    @Before
    override fun setUp() {
        super.setUp()

        validNpmPackageJson = File(FileOperation.getFileResource("valid-ankidroid-js-addon-test.json"))
        notValidNpmPackageJson = File(FileOperation.getFileResource("not-valid-ankidroid-js-addon-test.json"))
        addonsPackageListTestJson = FileOperation.getFileResource("test-js-addon.json")
    }

    @Test
    @Throws(IOException::class)
    fun isValidAnkiDroidAddonTest() {
        // test addon is valid or not, for a valid addon the result is Valid
        val result = getAddonModelFromJson(validNpmPackageJson)
        val addon = assertIs<AddonValidationResult.Valid>(result, "package.json contains required fields").addonModel

        // needs to test these fields
        assertEquals(addon.name, "valid-ankidroid-js-addon-test")
        assertEquals(addon.addonTitle, "Valid AnkiDroid JS Addon")
        assertEquals(addon.version, "1.0.0")
        assertEquals(addon.ankidroidJsApi, "0.0.3")
        assertEquals(addon.addonType, "reviewer")
        assertEquals(addon.icon, "") // reviewer icon is empty

        val expected: List<String> = listOf("ankidroid-js-addon")
        assertEquals(addon.keywords, expected)
    }

    @Test
    @Throws(IOException::class)
    fun notValidAnkiDroidAddonTest() {
        // test addon is valid or not, for a not valid addon the result is Invalid
        val result = getAddonModelFromJson(notValidNpmPackageJson)
        // assert that the package.json was not mapped to an addon model
        val errors = assertIs<AddonValidationResult.Invalid>(result, "package.json not contains required fields").errors
        // assert that error list contains error when the package.json not mapped to AddonModel
        assertFalse(errors.isEmpty())
    }

    @Test
    fun unreadableManifestReturnsErrorTest() {
        // a missing/unreadable package.json must be reported as Invalid, not thrown:
        // one corrupt addon directory must not abort listing the other addons
        val result = getAddonModelFromJson(File("/does/not/exist/package.json"))
        val errors = assertIs<AddonValidationResult.Invalid>(result, "model is not built from an unreadable manifest").errors
        assertFalse("unreadable manifest is reported as an error", errors.isEmpty())
    }

    @Test
    fun getAddonModelListFromJsonTest() {
        val url = File(addonsPackageListTestJson).toURI().toURL()
        val result = getAddonModelListFromJson(url)

        // first addon name and tgz download url
        val addon1 = result.first[0]
        assertEquals(addon1.name, "ankidroid-js-addon-progress-bar")
        assertThat(addon1.dist!!.tarball, endsWith(".tgz"))

        // second addon name and tgz download url
        val addon2 = result.first[1]
        assertEquals(addon2.name, "valid-ankidroid-js-addon-test")
        assertThat(addon2.dist!!.tarball, endsWith(".tgz"))
    }

    /**
     * A valid manifest, with individual fields overridable so each test can knock one out
     */
    private fun addonData(
        name: String? = "valid-ankidroid-js-addon-test",
        addonTitle: String? = "Valid AnkiDroid JS Addon",
        icon: String? = "",
        version: String? = "1.0.0",
        description: String? = "A test addon",
        main: String? = "index.js",
        ankidroidJsApi: String? = CURRENT_JS_API_VERSION,
        addonType: String? = REVIEWER_ADDON,
        keywords: List<String>? = listOf(ANKIDROID_JS_ADDON_KEYWORDS),
        author: Map<String, String>? = mapOf("name" to "AnkiDroid"),
        license: String? = "MIT",
        homepage: String? = "https://example.com",
        dist: DistInfo? = DistInfo("https://example.com/addon.tgz"),
    ): AddonData =
        AddonData(
            name,
            addonTitle,
            icon,
            version,
            description,
            main,
            ankidroidJsApi,
            addonType,
            keywords,
            author,
            license,
            homepage,
            dist,
        )

    @Test // the validator must report errors, never throw
    fun missingNameReturnsErrorTest() {
        val result = getAddonModelFromAddonData(addonData(name = null))
        val errors = assertIs<AddonValidationResult.Invalid>(result, "model is not built from an invalid manifest").errors
        assertFalse("missing 'name' is reported as an error", errors.isEmpty())
    }

    @Test // the validator must report errors, never throw
    fun missingKeywordsReturnsErrorTest() {
        val result = getAddonModelFromAddonData(addonData(keywords = null))
        val errors = assertIs<AddonValidationResult.Invalid>(result, "model is not built from an invalid manifest").errors
        assertFalse("missing 'keywords' is reported as an error", errors.isEmpty())
    }

    @Test // 'version' is required during model construction, so it must be validated
    fun missingVersionReturnsErrorTest() {
        val result = getAddonModelFromAddonData(addonData(version = null))
        val errors = assertIs<AddonValidationResult.Invalid>(result, "model is not built from an invalid manifest").errors
        assertFalse("missing 'version' is reported as an error", errors.isEmpty())
    }

    @Test // exercises the checks after the required-fields guard: their errors must be returned
    fun invalidAddonTypeReturnsErrorTest() {
        val result = getAddonModelFromAddonData(addonData(addonType = "invalid-type"))
        val errors = assertIs<AddonValidationResult.Invalid>(result, "model is not built from an invalid manifest").errors
        assertFalse("invalid 'addonType' is reported as an error", errors.isEmpty())
    }

    @Test
    fun missingDistIsValidTest() {
        // 'dist' is metadata added by the npm registry API; the package.json inside a
        // tarball does not contain it, so a locally installed addon must still validate
        val result = getAddonModelFromAddonData(addonData(dist = null))
        assertIs<AddonValidationResult.Valid>(result, "model is built from a manifest without 'dist'")
    }

    @Test
    fun missingOptionalMetadataIsValidTest() {
        // description/author/license are commonly absent from real package.json files
        val result = getAddonModelFromAddonData(addonData(description = null, author = null, license = null))
        val addon = assertIs<AddonValidationResult.Valid>(result, "model is built from a manifest without optional metadata").addonModel
        assertEquals("", addon.description)
        assertEquals(emptyMap<String, String>(), addon.author)
        assertEquals("", addon.license)
    }
}
